src/lib/sqle/connection.service.ts

```diff
@@ -1,74 +1,223 @@
+  username: string;
+}
+
+export interface IConnectOptions {
+  timeout: number;
+  attempts: number;
+}
+
+export function stringifyConnection(connection: ISQLEConnection): string {
+  if (connection) {
+    return `${connection.system.nickname}${connection.creds}`;
   }
 }
 
 @Injectable()
 export class VantageConnectionService {
-  constructor(private _queryService: VantageQueryService) {}
+  private readonly _currentConnectionSubject: BehaviorSubject<ISQLEConnection> = new BehaviorSubject<ISQLEConnection>(
+    this._getConnectionState().current,
+  );
+  private readonly _connectionsSubject: BehaviorSubject<ISQLEConnection[]> = new BehaviorSubject<ISQLEConnection[]>(
+    this._getConnectionState().connections,
+  );
+  public currentConnection$: Observable<ISQLEConnection> = this._currentConnectionSubject.asObservable();
+  public connections$: Observable<ISQLEConnection[]> = this._connectionsSubject.asObservable();
+
+  constructor(private _queryService: VantageQueryService, private _sessionService: VantageSessionService) {
+    const connectionState: IVantageConnectionState = this._getConnectionState();
+    if (connectionState && connectionState.username !== this._currentUsername) {
+      // mismatch, so clear
+      this._connections = [];
+      this._currentConnection = undefined;
+    }
+  }
+
+  private set _currentConnection(connection: ISQLEConnection) {
+    this._setConnectionState({
+      current: connection,
+      connections: this._connections,
+      username: this._currentUsername,
+    });
+    this._currentConnectionSubject.next(connection);
+  }
+  private get _currentConnection(): ISQLEConnection {
+    return this._currentConnectionSubject.getValue();
+  }
+  public get currentConnection(): ISQLEConnection {
+    return this._currentConnection;
+  }
+
+  private set _connections(connections: ISQLEConnection[]) {
+    this._setConnectionState({
+      current: this._currentConnection,
+      connections,
+      username: this._currentUsername,
+    });
+    this._connectionsSubject.next(connections);
+  }
+  private get _connections(): ISQLEConnection[] {
+    return this._connectionsSubject.getValue();
+  }
+  public get connections(): ISQLEConnection[] {
+    return this._connections;
+  }
+
+  public addConnectionAndConnect(connection: ISQLEConnection, options?: IConnectOptions): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      throw Error('Connection already exists');
+    } else {
+      return this._pingAndSave(connection, true, options);
+    }
+  }
+
+  public addConnection(connection: ISQLEConnection, options?: IConnectOptions): Observable<ISQLEConnection> {
+    if (this._getConnectionIndex(connection) > -1) {
+      throw Error('Connection already exists');
+    } else {
+      return this._pingAndSave(connection, false, options);
+    }
+  }
 
-  public get current(): ISQLEConnection {
-    return current();
+  public connectToExistingConnection(
```

Comment:
Whats the usage of `connecting` to an existing connection? Shouldnt that never happen since you always just connect to a new connection?

Reply:
So this is used when switching between connections that already have been created. So the connection is already in the connections array, but it is just not the current connection.